fix inclusion of PRO in secondary structure

orbeckst opened the pull request:
Fixes #4913

Changes made in this Pull Request:
<!-- Describe the changes that this PR makes. If applicable, use the following bullet list. --> 
- ported fix from PyDSSP 0.9.1 by @ShintaroMinami to analysis.dssp.DSSP (see also https://github.com/ShintaroMinami/PyDSSP/issues/2)
- new kwarg ignore_proline_donor=True for DSSP (the new default changes the behavior and implements the fix, False recovers old behavior); the kwarg also exists in PyDSSP
- updated docs
- minimal regression tests
- updated CHANGELOG

## PR Checklist
<!-- Please use the following checklist to ensure the PR is ready to be reviewed/merged. -->
 - [x] Issue raised/referenced?
 - [x] Tests updated/added?
 - [x] Documentation updated/added?
 - [x] `package/CHANGELOG` file updated?
 - [x] Is your name in `package/AUTHORS`? (If it is not, add it!)

## Developers Certificate of Origin
<!-- 
In order for this PR to be merged you **must certify that you are able to submit your code to be included in MDAnalysis**.
-->

I certify that I can submit this code contribution as described in the [**Developer Certificate of Origin**](https://developercertificate.org/), under the MDAnalysis [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE).


<!-- readthedocs-preview mdanalysis start -->
----
📚 Documentation preview 📚: https://mdanalysis--5065.org.readthedocs.build/en/5065/

<!-- readthedocs-preview mdanalysis end -->

orbeckst committed:
fix inclusion of PRO in secondary structure

- fix #4913
- ported fix from PyDSSP 0.9.1 by @ShintaroMinami to analysis.dssp.DSSP
  (see also https://github.com/ShintaroMinami/PyDSSP/issues/2)
- new kwarg ignore_proline_donor=True for DSSP (the new default changes the behavior
  and implements the fix, False recovers old behavior); the kwarg also exists in
  PyDSSP
- updated docs
- minimal regression tests
- updated CHANGELOG

orbeckst commented on package/MDAnalysis/analysis/dssp/dssp.py:
```diff
@@ -315,6 +329,9 @@ def __init__(
             ]
             for t in heavyatom_names
         }
+        self._donor_mask: Optional[np.ndarray] = (
+            ag.residues.resnames != "PRO" if ignore_proline_donor else None
+        )
```
This may not be correct. The code runs ... but I am not sure if I should be masking corresponding atoms.

marinegor replied:
if the question is about shape of the donor mask (i.e. make it for residues or for atoms), then it seems to be correct -- pydssp makes it of shape `l` ([line](https://github.com/ShintaroMinami/PyDSSP/blob/e251a43ff8622fe0a555313b1567edce45e789e8/pydssp/pydssp_numpy.py#L66)) and `l` is the number of residues ([link](https://github.com/ShintaroMinami/PyDSSP/blob/e251a43ff8622fe0a555313b1567edce45e789e8/pydssp/pydssp_numpy.py#L45)). However, the notation I made in `pydssp_numpy.py` is wrong, hence I renamed it in the commit below.

orbeckst commented on package/MDAnalysis/analysis/dssp/pydssp_numpy.py:
```diff
@@ -130,6 +131,12 @@ def get_hbond_map(
         (without or with hydrogens). If hydrogens are not present, then
         ideal positions (see :func:_get_hydrogen_atom_positions) are used.
+    donor_mask : np.array
+         Mask out any hydrogens that should not be considered (in particular HN
+         in PRO). If ``None`` then all H will be used (behavior up to 2.9.0).
```
These docs should be more specific and state the shape. I just quickly guessed the shape from https://github.com/ShintaroMinami/PyDSSP/blob/e251a43ff8622fe0a555313b1567edce45e789e8/scripts/pydssp#L30
```python
donor_mask = sequence != 'PRO' if args.ignore_proline_donor else None
```

marinegor replied:
Yep, you're right -- and the name `n_atoms` in the function below was wrong, which I guess has caused your confusion. I've added some docs here as well.

orbeckst commented on package/MDAnalysis/analysis/dssp/pydssp_numpy.py:
```diff
@@ -194,15 +205,26 @@ def get_hbond_map(
+        if donor_mask is not None
+        else np.ones(n_atoms, dtype=float)
+    )
+    donor_mask = np.tile(donor_mask[:, np.newaxis], (1, n_atoms))
```
Is the donor_mask (one element for each residue) really correct for this tiling????

orbeckst commented on package/MDAnalysis/analysis/dssp/pydssp_numpy.py:
```diff
@@ -194,15 +205,26 @@ def get_hbond_map(
     hbond_map = (np.sin(hbond_map / margin * np.pi / 2) + 1.0) / 2
-    hbond_map = hbond_map * local_mask
+    hbond_map *= local_mask
+    hbond_map *= donor_mask
```
Is this correct? The original code uses https://github.com/ShintaroMinami/PyDSSP/blob/e251a43ff8622fe0a555313b1567edce45e789e8/pydssp/pydssp_numpy.py#L72
```python
hbond_map = hbond_map * repeat(donor_mask, 'l1 l2 -> b l1 l2', b=b)
```
(with `einops.repeat()`). Note that we create our `donor_mask` with `tile` so it may already be the right size and shape.

marinegor replied:
yep, they're all the same shape internally.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_dssp.py:
```diff
@@ -13,15 +13,52 @@
 def test_file_guess_hydrogens(pdb_filename, client_DSSP):
+    # run 2.9.0 tests (which include PRO)
+    # ignore_proline_donor=False
+    # TODO: update reference data for ignore_proline_donor=True
```
We should really have correct reference data. About half of the files do not show a difference between `ignore_proline_donor=True` and `ignore_proline_donor=False`.

marinegor replied:
after some while, I figured what's the problem: the updated pydssp just gives different results now! Namely, take ` 3e8mA.pdb`:

```bash
0.9.0: -----EEEE-----------EE-----EE-----HHHHHHHHHHH---EEEEE-----HHHHHHHHH----EEE-----HHHHHHHHHHHH---HHHEEEE---HHHHHHH----EEE------HHHH--------------HHHHHHHHH----HHHHHHH-- 3e8mA.pdb
0.9.1: -----EEEE-----------EE-----EE-----HHHHHHHHHHH----EEEE-----HHHHHHHHH----EEE-----HHHHHHHHHHHH---HHHEEEE---HHHHHHH----EEE------HHHH--------------HHHHHHHHH----HHHHHHH--
                                                       X
``

In this particular case it used to (incorrectly) assign `E` (sheet) char to P48, and now it correctly assigns it to `-` (loop). The author didn't catch it since their test checks only correlation between pydssp and original assignment, which is still high enough since 0.9.0 changes only PRO assignment (and not even always).

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_dssp.py:
```diff
@@ -32,8 +69,6 @@ def test_trajectory(client_DSSP):
     assert (
         first_frame[:10] != last_frame[:10] == avg_frame[:10] == "-EEEEEE---"
     )
-    protein = mda.Universe(TPR, XTC).select_atoms("protein")
-    run = DSSP(protein).run(**client_DSSP, stop=10)
```
These lines seemed superfluous as the atomgroup approach is tested separately.

marinegor replied:
agree

orbeckst reviewed (commented):
This is a quick draft. I'd be more than happy if someone continued and completed it.

(EDIT: ... and I am also grateful for any comments that would help me to continue working on it.)

orbeckst committed:
Merge branch 'develop' into update-dssp-proline-fix

orbeckst committed:
Merge branch 'develop' into update-dssp-proline-fix

orbeckst committed:
Merge branch 'develop' into update-dssp-proline-fix

orbeckst committed:
Merge branch 'develop' into update-dssp-proline-fix

orbeckst commented:
@marinegor if you can spare a moment to look at this PR then that would be great; I think you're really the person with best insight into the DSSP code. Feel free to tell me that all these changes are rubbish and we need to redo everything.

marinegor commented:
@orbeckst I don't think it's all rubbish -- when this change was introduced to pydssp, I also made a similar attempt of introducing it to MDAnalysis.

My main concern is how to combine it with a workaround I introduced when PRO fix wasn't yet available. I have a feeling that in principle we should just get rid of that (I mean [roughly this part](https://github.com/MDAnalysis/mdanalysis/pull/5065/files#r2141302741)) and rely purely on pydssp. I'll have a closer look.

marinegor committed:
fix variable names and update dssp files after pydssp 0.9.1
  - fixed variable names from `n_atoms` to `n_residues`
  - updated some docs, and most importantly
  - added updated `dssp` files.

marinegor reviewed (changes requested):
Quick summary: pydssp in 0.9.1 fixed the proline bug and changed (in correct direction) assignment of prolines. For example, for P48 in  3e8mA.pdb assignment used to be `E`, and now it's correctly `-`. Also the [dssp output file](https://github.com/ShintaroMinami/PyDSSP/blob/master/tests/testset/TS50/dssp/3e8mA.dssp) in the original repo annotates P48 as `-`.

I suggest in this PR we:
 - change test files following up the pydssp update
 - remove the `ignore_proline_donor` option in function signature (i.e. make it always `True`, since the previous implementation was plain wrong)

@orbeckst while reviewing the PR, I made few commits on top of your branch -- namely, fixed variable names from `n_atoms` to `n_residues`, updated some docs, and most importantly, added updated `dssp` files. They're in here, feel free to pull them for yourself: //github.com/marinegor/mdanalysis/pull/new/update-dssp-proline-fix

orbeckst commented:
@marinegor thank you for the detailed review. Could you please directly update this PR with your changes? I’d really appreciate that! 

I’m not sure if I get to working on the PR before the UGM this weekend. Please feel free to make any changes necessary!

thanks again !!

marinegor committed:
remove ignore_proline_donors

marinegor committed:
add myself to the changelog

marinegor commented:
@orbeckst sure, I've pushed them on top of this branch.

orbeckst commented:
Thanks!!


> On Nov 5, 2025, at 04:31, Egor Marin ***@***.***> wrote:
> 
> 
> marinegor
>  left a comment 
> (MDAnalysis/mdanalysis#5065)
>  <https://github.com/MDAnalysis/mdanalysis/pull/5065#issuecomment-3490684995>
> @orbeckst <https://github.com/orbeckst> sure, I've pushed them on top of this branch.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/MDAnalysis/mdanalysis/pull/5065#issuecomment-3490684995>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAB2DHGLGDX4FYVKHCRCBQL33HNX7AVCNFSM6AAAAAB7D57EXGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTIOJQGY4DIOJZGU>.
> You are receiving this because you were mentioned.
> 

--
Oliver Beckstein (he/his/him)
***@***.***

orbeckst commented on package/MDAnalysis/analysis/dssp/pydssp_numpy.py:
```diff
@@ -128,8 +129,15 @@ def get_hbond_map(
+        are used.
+    donor_mask : np.array
+         Mask out any hydrogens that should not be considered (in particular HN
+         in PRO). If ``None`` then all H will be used (behavior up to 2.9.0).
```
```suggestion
         in PRO). If ``None`` then all H will be used (behavior up to 2.10.0).
```

orbeckst reviewed (commented):
Thank you for pushing the PR across the finish line @marinegor !

I'll just push a minor doc fix and then we can squash-merge. Please feel free to do the honors. When I squash-merge I typically clean up the log and format it something like
```
title

- fix #xxx
- summary of changes (typically from PR summary) and/or CHANGELOG
- other important changes (from comments)
- tests updated
- docs updated
- CHANGELOG updated

---
lines identifying code contributors as provided by GH
```

orbeckst committed:
doc fix with release number

marinegor commented:
@orbeckst thanks, did my best! Initiation PR, I guess🥳

orbeckst commented:
Awesome, great work @marinegor !
